### PR TITLE
JENKINS-69268-fix-bitbucket-shared-library-build-status

### PR DESCRIPTION
**Pull request by djansons-atlassian**
Fixes the bug where the build status is not sent correctly when global pipeline libraries are used in a job.

https://issues.jenkins.io/browse/JENKINS-69268?filter=-1

This change ensures that build status is sent to all repos used in the job, except for repos hosting a global pipeline library.

I still need to perform testing with a freestyle job and a pipeline job and write extra unit tests.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [] Ensure that the pull request title represents the desired changelog entry - WIP PR - N/A 
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/status/LocalSCMListener.java [hunk 1]
```diff
@@ -37,6 +39,17 @@ public LocalSCMListener() {
     public void onCheckout(Run<?, ?> build, SCM scm, FilePath workspace, TaskListener listener,
                            @CheckForNull File changelogFile,
                            @CheckForNull SCMRevisionState pollingBaseline) {
+
+        for (LibraryResolver resolver : ExtensionList.lookup(LibraryResolver.class)) {
```
[the remark this answers is not in the document]
Comment (djansons-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/status/LocalSCMListener.java:61]: Addressed all of the above.

**Comment by djansons-atlassian** [comment]
I tried my best to mock Jenkins with the JenkinsProvider, but ran into issues with the mock ExtentionList. Instead I am using JenkinsRule for the unit tests. 
In addition to the extra unit tests, I performed the following manual testing for pipeline and multi branch pipeline jobs:

- status is not sent to library SCMs
- status is sent to non library SCMS
- status is sent to multiple non Library SCMs
- status is sent to a library SCM if there is a bbs_checkout step for the repo that hosts the library SCM. 

I did not get around to writing integration tests. Please advise if these are required.

**Review by dkjellin — APPROVED** [review]

**Comment by dkjellin** [comment]
> I did not get around to writing integration tests. Are these required, or does the unit test suffice?

Generally I would prefer that, but given you use the Jenkins rule I think your tests are probably enough

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/status/BuildStatusPosterTest.java [hunk 2]
```diff
@@ -15,6 +15,9 @@
 import org.mockito.junit.MockitoJUnitRunner.Silent;
```
Comment (mhenschke-atl) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/status/BuildStatusPosterTest.java:15]: We should add a test for this class testing when multiple revision actions are present on a run, to ensure all are having their restriction sent through correctly.

**Comment by mhenschke-atl** [comment]
I think we should have an integration test for this before it merges.`BitbucketSCMIT` appears the most appropriate place to add such a test- creating a job with a shared library, we have some rest tests already verifying build statuses are posting, so I think it's worthwhile creating a new test in that suite with a shared library to ensure the attached repo does not have a build status against it.

**Comment by djansons-atlassian** [comment]
> I think we should have an integration test for this before it merges.`BitbucketSCMIT` appears the most appropriate place to add such a test- creating a job with a shared library, we have some rest tests already verifying build statuses are posting, so I think it's worthwhile creating a new test in that suite with a shared library to ensure the attached repo does not have a build status against it.

I have refactored the code and written some unit tests for global libraries and shared folder libraries. Hopefully this suffices instead of the integration tests.

**Review by dkjellin — APPROVED** [review]

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/status/LocalSCMListenerTest.java [hunk 3]
```diff
@@ -62,25 +75,27 @@ public void setup() throws URISyntaxException {
...
+        LibraryConfiguration libConfig = mock(LibraryConfiguration.class);
+        SCMRetriever scmRetriever = mock(SCMRetriever.class);
+        when(libConfig.getRetriever()).thenReturn(scmRetriever);
+        when(scmRetriever.getScm()).thenReturn(bitbucketSCM);
+        FolderLibraries folderLibraries = new FolderLibraries(singletonList(libConfig));
+        workflowMultiBranchProject.addProperty(folderLibraries);
+
+        listener.onCheckout(run, bitbucketSCM, null, taskListener, null, null);
+
+        verify(bitbucketSCM, times(2)).getId();
```
Comment (mhenschke-atl) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/status/LocalSCMListenerTest.java:98]: Why is this exactly two? If this is just part of the flow of the test and you just want to verify some number, I think `atLeast(1)` might be better so it doesn't break tests in the future, but if that number has a specific meaning here then times(2) should be fine

**Review by mhenschke-atl — APPROVED** [review]

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/status/LocalSCMListenerTest.java (hunk 3 above)
Comment (djansons-atlassian) [reply src/test/java/com/atlassian/bitbucket/jenkins/internal/status/LocalSCMListenerTest.java:98]: On line 165 I put a comment about why it's 2. I have added comments on all the other `times(2)`s to add some clarity.

**Pull request merged**